bjerk-style.md

```diff
@@ -0,0 +1,411 @@
+
+* If you don't mean a function argument to be copied when passed by value, and if the argument type
+  is more than 16 bytes, then pass the argument as `*const`. This will catch bugs where the caller
+  makes an accidental copy on the stack before calling the function.
+
+* **Shrink the scope** to minimize the number of variables at play and reduce the probability that
+  the wrong variable is used.
+
+* Calculate or check variables close to where/when they are used. **Don't introduce variables before
+  they are needed.** Don't leave them around where they are not. This will reduce the probability of
+  a POCPOU (place-of-check to place-of-use), a distant cousin to the infamous
+  [TOCTOU](https://en.wikipedia.org/wiki/Time-of-check_to_time-of-use). Most bugs come down to a
+  semantic gap, caused by a gap in time or space, because it's harder to check code that's not
+  contained along those dimensions.
+
+* Use simpler function signatures and return types to reduce dimensionality at the call site, the
+  number of branches that need to be handled at the call site, because this dimensionality can also
+  be viral, propagating through the call chain. For example, as a return type, `void` trumps `bool`,
+  `bool` trumps `u64`, `u64` trumps `?u64`, and `?u64` trumps `!u64`.
+
+* Ensure that functions run to completion without suspending, so that precondition assertions are
+  true throughout the lifetime of the function. These assertions are useful documentation without a
+  suspend, but may be misleading otherwise.
+
+* Be on your guard for **[buffer bleeds](https://en.wikipedia.org/wiki/Heartbleed)**. This is a
+  buffer underflow, the opposite of a buffer overflow, where a buffer is not fully utilized, with
+  padding not zeroed correctly. This may not only leak sensitive information, but may cause
+  deterministic guarantees as required by TigerBeetle to be violated.
+
+* Use newlines to **group resource allocation and deallocation**, i.e. before the resource
+  allocation and after the corresponding `defer` statement, to make leaks easier to spot.
+
+### Off-by-one errors
+
+* **The usual suspects for off-by-one errors are casual interactions between an `index`, a `count`
+  or a `size`.** These are all primitive integer types, but should be seen as distinct types, with
+  clear rules to cast between them. To go from an `index` to a `count` you need to add one, since
+  indexes are *0-based* but counts are *1-based*. To go from a `count` to a `size` you need to
+  multiply by the unit. Again, this is why including units and qualifiers in variable names is
+  important.
+
+* Show your intent with respect to division. For example, use `@divExact()`, `@divFloor()` or
+  `div_ceil()` to show the reader you've thought through all the interesting scenarios where
+  rounding may be involved.
+
+### Style by the numbers
+
+* Run `zig fmt`.
+
+* Use 4 spaces of indentation, rather than 2 spaces, as that is more obvious to the eye at a
```

Comment:
We typically use 2 spaces rather than 4, no? should that be changed?

Reply:
Good question. It is a change, yes. What are you thoughts on it?

Reply:
I think I'm coming around to 4 spaces. We don't need to squeeze the code together.